pkg/agent/controller/serviceimport.go

```diff
@@ -0,0 +1,217 @@
+	if err != nil {
+		return nil, fmt.Errorf("Error building lighthouseClient %s", err.Error())
+	}
+
+	serviceImportController := ServiceImportController{
+		queue:            workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		kubeClientSet:    kubeClientSet,
+		lighthouseClient: lighthouseClient,
+		clusterID:        spec.ClusterID,
+		namespace:        spec.Namespace,
+	}
+
+	return &serviceImportController, nil
+}
+
+func (c *ServiceImportController) Start(stopCh <-chan struct{}) error {
+	informerFactory := externalversions.NewSharedInformerFactoryWithOptions(c.lighthouseClient, 0,
+		externalversions.WithNamespace(c.namespace))
+	c.serviceInformer = informerFactory.Lighthouse().V2alpha1().ServiceImports().Informer()
+
+	c.serviceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			klog.V(2).Infof("ServiceImport %q added", key)
+			if err == nil {
+				c.queue.Add(key)
+			}
+		},
+		UpdateFunc: func(obj interface{}, new interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(new)
+			// TODO Change level
+			klog.Infof("ServiceImport %q updated", key)
+			if err == nil {
+				c.queue.Add(key)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			klog.Infof("ServiceImport %q deleted", key)
+			if err == nil {
+				var si *lighthousev2a1.ServiceImport
+				var ok bool
+				if si, ok = obj.(*lighthousev2a1.ServiceImport); !ok {
+					tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+					if !ok {
+						klog.Errorf("Failed to get deleted serviceimport object for key %s, serviceImport %v", key, si)
+						return
+					}
+
+					si, ok = tombstone.Obj.(*lighthousev2a1.ServiceImport)
+
+					if !ok {
+						klog.Errorf("Failed to convert deleted tombstone object %v  to serviceimport", tombstone.Obj)
+						return
+					}
+				}
+				if si.Spec.Type != lighthousev2a1.Headless {
+					return
+				}
+				c.serviceImportDeletedMap.Store(key, si)
+				c.queue.AddRateLimited(key)
+			}
+		},
+	})
+
+	go c.serviceInformer.Run(stopCh)
+	go c.runServiceImportWorker()
+
+	go func(stopCh <-chan struct{}) {
+		<-stopCh
+		c.queue.ShutDown()
+
+		klog.Infof("ServiceImport Controller stopped")
+	}(stopCh)
+
+	return nil
+}
+
+func (c *ServiceImportController) runServiceImportWorker() {
+	for {
+		keyObj, shutdown := c.queue.Get()
+		if shutdown {
+			klog.Infof("Lighthouse watcher for ServiceImports stopped")
+			return
+		}
+
+		key := keyObj.(string)
+
+		func() {
+			defer c.queue.Done(key)
+			obj, exists, err := c.serviceInformer.GetIndexer().GetByKey(key)
+
+			if err != nil {
+				klog.Errorf("Error retrieving the object with store is  %v from the cache: %v", c.serviceInformer.GetIndexer().ListKeys(), err)
+				// requeue the item to work on later
+				c.queue.AddRateLimited(key)
+
+				return
+			}
+
+			c.queue.Forget(key)
```

Comment:
This is a bit premature as it could get re-queued below. I suggest structuring this like you did in the endpoint controller, ie returning an error from the functions below etc

Reply:
Done.